cmd/signing-server/main.go

```diff
@@ -69,14 +73,126 @@ type Config struct {
 	DisableAuth        bool
 	DisableHTTPS       bool
 
+	HSMModule   string
+	HSMSlot     int
+	HSMPass     string
+	HSMKeyLabel string
+	HSMKeyId    string
+
+	HSMContext   *pkcs11.Ctx
+	HSMSession   pkcs11.SessionHandle
+	HSMKeyHandle pkcs11.ObjectHandle
+
 	// calculated by program
 	Logger *zap.Logger
 }
 
+func (c *Config) SetupHSM() error {
+	c.Logger.Info("setup HSM signing", zap.String("module", c.HSMModule))
+	p := pkcs11.New(c.HSMModule)
+	err := p.Initialize()
+	if err != nil {
+		return err
+	}
+
+	if c.HSMSlot < 0 {
+		slots, err := p.GetSlotList(true)
+		if err != nil {
+			return fmt.Errorf("lookup HSM slots: %w", err)
+		} else {
+			c.HSMSlot = int(slots[0])
+		}
+	}
+	c.Logger.Info("using slot", zap.Int("slot", c.HSMSlot))
+
+	session, err := p.OpenSession(uint(c.HSMSlot), pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	if err != nil {
+		return fmt.Errorf("open HSM session: %w", err)
+	}
+
+	err = p.Login(session, pkcs11.CKU_USER, c.HSMPass)
+	if err != nil {
+		p.CloseSession(session)
```

Comment:
CloseSession's errors are ignored everywhere.

Reply:
the session is always closed, here, because of an error, which exits the program anyway. Reporting the session close error does therefore not provide any value to the caller.